### ISSUE TITLE
How to refresh display of Sublime Linter when a file is changed

### ISSUE THREAD

**Issue by pierresh** [SublimeLinter/SublimeLinter#1963]
I am using the plugin [SublimeLinter-contrib-phpstan](https://github.com/Rockstar04/SublimeLinter-contrib-phpstan), which works well, but I only see linting changes when I save the file. I would like to see the linting in real-time, so I want it to happen whenever I make a change.

To achieve this, I removed the setting `tempfile_suffix = "-"`, then the plugin is executed each time I make a change to my PHP file (I verified this by adding a simple `print('here')` on [the line 54](https://github.com/Rockstar04/SublimeLinter-contrib-phpstan/blob/master/linter.py#L54), just before the return). However, the linting results are not displayed by Sublime Linter, when it works perfectly when I save the file.

What should I do to get the linting results displayed in Sublime Text when the file is changed? Thanks.

**Comment by kaste** [comment]
The plugin is not only executed after each key press it also sends the view's content via stdin to the linter.  The linter therefore must accept this streaming in.  You typically need to change how you invoke the linter (the `cmd`) as well.  Unsure if phpstan even supports this.  Maybe `tempfile_suffix = "php"` t.i. working with temporary files is an option, but only if the actual file names and locations don't matter.

I noticed that you implement `get_cmd`.  I don't recommend that as it is not an extension point and I could change the implementation here.  It is also not clear why you need to do that.  Actually `cmd()` should never return `[]`, it returns a list/tuple of strings or raises to abort.  Typically after a `self.notify_failure()` you `raise PermanentError("options mismatch or whatever the reason is in plain English")`.

`enable_cells` may be a misuse as well.  Not sure.  The other php plugins use `embedding.php, source.php`, maybe that would work here too.  A "cell" is meant to be something in a notebook.  Like in literal programming.  Code blocks (=cells) within prose.  Like in any markdown file.

**Comment by pierresh** [comment]
Thanks a lot for your feedback. But still I do not understand why the results can be shown only when saving files, not when making changes to the file. The plugin is called exactly with the same parameters, especially `--error-format=raw` to output one error per line, with path to file, line number, and error description.

I took over this plugin recently and your comment about `get_cmd` leaded me to review completely [this plugin](https://github.com/Rockstar04/SublimeLinter-contrib-phpstan/blob/master/linter.py) and I figured that actually most of the code could be removed :)

About `enable_cells`: Sublime Linter 4.26 broke that plugin and I needed to activate `enable_cells` to make it working again.  Indeed `embedding.php, source.php` also restores it but, as PHPStan focuses on PHP classes and not files mixing PHP and HTML tags, I supposed it would be cleaner to stay with `source.php`.

**Comment by pierresh** [comment]
Actually, I found that there is a `debug` mode in Sublime Linter, which helped me to understand what is happening: PHPStan analyzes the file on the disk (which is not saved yet) thus still finding again the previous results, when the file was saved..
The solution should be something like providing content of the file to PHPStan, instead of the path to the file being edited.

**Comment by kaste** [comment]
That is what I was trying to say.  I found https://github.com/phpstan/phpstan/issues/3262 which is closed but not implemented.  

About `enable_cells`, that's just a semantic hint.  The idea is that a the plugin works as is for all the typical source code variants which are php and html files I guess in this case.  But it should be opt-in for the literal coding or documentation files, typically these are markdown files, but there are other file types as well.
